router/core/header_rule_engine.go

```diff
@@ -399,11 +416,21 @@ func (h *HeaderPropagation) applyRequestRule(ctx RequestContext, request *http.R
 		}
 
 		if rule.Expression != "" {
-			value, err := h.getRequestRuleExpressionValue(rule, reqCtx)
+			// If reqCtx is nil we still don't want to add the header because there is an expression
+			if reqCtx == nil {
+				return
+			}
+			value, err := h.getRequestRuleExpressionValue(rule, reqCtx.expressionContext, rule.Operation)
 			if err != nil {
+				reqCtx.logger.Error("error occurred while getting expr val", zap.Error(err))
 				reqCtx.SetError(err)
-			} else if value != "" {
-				request.Header.Set(rule.Name, value)
+			} else {
+				stringVal, ok := value.(string)
+				if !ok {
+					reqCtx.logger.Error("unexpected non string value", zap.Any("value", value))
```

Comment:
I don't think that the user will get enough context to understand and investigate in this error. We should have a scoped logger.